Add support of Double Ratchet (PFS) chats

SergeySeroshtan opened the pull request.

Ogerets committed:
Cherry-pick commit: Updated dependencies, fixed VirgilAuthorizer

Cherry-pick commit: 663f7352ca753f9bc64f07996cb5fc1d06c32840

# Conflicts:
#	Notification Extension/NotificationService.swift
#	Source/Helpers/Virgil/LocalKeyManager.swift
#	Source/Helpers/Virgil/VirgilAuthorizer.swift
#	VirgilMessenger.xcodeproj/project.pbxproj

SergeySeroshtan committed:
Add support of Double Ratchet (PFS) chats

Ogerets commented on Source/Helpers/MessageProcessor.swift:
```diff
@@ -15,19 +16,19 @@ class MessageProcessor {
         var decryptedAdditional: Data?
 
         if let data = encryptedMessage.additionalData {
             decryptedAdditional = try Virgil.ethree.authDecrypt(data: data, from: channel.getCard())
```
Here should be option with ratchet decryption. Also, this decryption, likely, should be moved to MessageProcessor.decrypt() somehow

Ogerets commented on Source/Helpers/MessageProcessor.swift:
```diff
@@ -73,6 +74,14 @@ class MessageProcessor {
+        case .newChannel(let newChannel):
+            if newChannel.type == .singleRatchet {
+                _ = try Virgil.ethree.joinRatchetChannel(with: channel.getCard()).startSync().get()
+                channel.type = .singleRatchet
```
I think we should change such things using Storage class. In this way we wrap storage interactions (saveContext) in one place

Ogerets commented on Source/Helpers/MessageSender.swift:
```diff
@@ -15,22 +16,51 @@ public class MessageSender {
+        try Ejabberd.shared.send(encryptedMessage, to: channel.name)
+    }
+
+    private func encrypt(data: Data, for card: Card, withRacthetChannel ratchetChannel: RatchetChannel?) throws -> Data {
```
withRatchetChannel*
And is swift style it should be just `with:`, since we have specific parameter type

Ogerets commented on Source/Helpers/MessageSender.swift:
```diff
@@ -15,22 +16,51 @@ public class MessageSender {
 
-        let card = try channel.getCard()
-        let ciphertext = try Virgil.ethree.authEncrypt(data: exported, for: card)
+        let ciphertext = try self.encrypt(data: exported, for: card, withRacthetChannel: ratchetChannel)
```
we can pass here channel type to remove constructing RatchetChannel above and encapsulate Virgil work on encryption in one func. Also, we can consider moving all wrapped encrypt decrypt methods to Virgil class, since it's a wrapper itself

Ogerets commented on Source/Helpers/MessageSender.swift:
```diff
@@ -159,4 +189,18 @@ public class MessageSender {
         }
     }
+
+    public func send(newChannel: Message.NewChannel, date: Date, channel: Storage.Channel, completion: @escaping (Error?) -> Void) {
```
Why do we need this func? We can just use sync sendService from startRatchetChannel

SergeySeroshtan replied:
Function become **public**:
```swift
MessageSender.public func sendService(message: Message...)
```

Ogerets commented on Source/Models/ChannelType.swift:
```diff
@@ -0,0 +1,15 @@
+
+public enum ChannelType: String, Codable {
+    case single
+    case group
```
Maybe lets remove types we not support yet? Since e.g. it makes us add this case to every switch.

Ogerets commented on Source/Storage/Storage+Channel.swift:
```diff
@@ -135,7 +130,7 @@ extension Storage {
 
         public func getCard() throws -> Card {
-            guard self.type == .single, let card = self.cards.first else {
+            guard (self.type == .single) || (self.type == .singleRatchet), let card = self.cards.first else {
```
Let's use switch here for channel.type. Otherwise, we'll miss this place someday

Ogerets commented on Source/Utils/Notifications.swift:
```diff
@@ -110,6 +110,9 @@ extension Notifications {
             let userInfo = [NotificationKeys.message.rawValue: iceCandidate]
             self.center.post(name: notification, object: self, userInfo: userInfo)
+
+        case .newChannel:
```
Redundant. caller of this method could directly call post(.chatListUpdated), since it has switch anyway

Ogerets commented:
Also, there's still an issue, when users create DR channel with each other simultaneously

Ogerets commented:
Don't forget, we need push notifications rework for DR

SergeySeroshtan committed:
Apply review comments to the Double Ratchet chats

SergeySeroshtan committed:
Remove enum value ChannelType.group

SergeySeroshtan committed:
swiftlint autocorrect

SergeySeroshtan committed:
Fix applied review comments to the Double Ratchet chats

SergeySeroshtan committed:
Solve race condition when two Double Ratchet channels are created near the same time

SergeySeroshtan commented on Source/Helpers/MessageProcessor.swift:
```diff
@@ -138,15 +128,29 @@ class MessageProcessor {
+                return try ratchetChannel.decrypt(data: data)
+            } catch {
+                return try Virgil.ethree.authDecrypt(data: data, from: channel.getCard())
+            }
```
@Ogerets This ugly solution is needed to handle a case when two Double Ratchet channels are created near the same time to be able to decrypt a Message.NewMessage message.

SergeySeroshtan committed:
Merge branch 'develop-voice' into develop-voice-ratchet

# Conflicts:
#	Notification Extension/NotificationService.swift
#	Source/Helpers/MessageProcessor.swift
#	Source/Helpers/MessageSender.swift
#	Source/Helpers/Virgil/Virgil.swift
#	Source/Storage/Storage+Channel.swift
#	Source/Storage/Storage+ChannelManagement.swift
#	VirgilMessenger.xcodeproj/project.pbxproj

SergeySeroshtan committed:
Remove Firebase.framework from the link libraries

SergeySeroshtan committed:
Add required frameworks to the extension targets

SergeySeroshtan committed:
Drop dependency to the Starscream.framework

SergeySeroshtan committed:
Update dependencies: FirebaseAnalyticsBinary, virgil-e3kit-x

SergeySeroshtan committed:
Add file Virgil+CryptoOpertaions.swift

SergeySeroshtan committed:
Extract encrypt/decrypt operations to the Virgil extension

SergeySeroshtan committed:
Update dependency: virgil-e3kit-x

SergeySeroshtan committed:
Use shared container within E3Kit

SergeySeroshtan committed:
Add new file Storage+MessageUI

SergeySeroshtan committed:
! Changed the way messages are exported to their UI representations

Motivation: Eliminate dependency to the Chatto Framework from the Notification Extension

SergeySeroshtan committed:
! Add Storage and all its derived classes to the notification extension

SergeySeroshtan committed:
Add CoreData.framework to the notification extensions

SergeySeroshtan committed:
Add account load when notification comes
